### PR TITLE
Add support for Libraries API

### PR DESCRIPTION
**Pull request by andrewmchen**
https://docs.databricks.com/api/latest/libraries.html

### CODE REVIEW

File: databricks_cli/libraries/cli.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+
+
+INSTALL_OPTIONS = ['jar', 'egg', 'maven-coordinates', 'pypi-package', 'cran-package']
+JAR_HELP = 'JAR uploaded to dbfs or s3.'
+EGG_HELP = 'Egg uploaded to dbfs or s3.'
+MAVEN_COORDINATES_HELP = """
+Maven coordinates in the form of GroupId:ArtifactId:Version (i.e. org.jsoup:jsoup:1.7.2).
+"""
+MAVEN_REPO_HELP = """
+Maven repo to install the Maven package from. If omitted, both Maven
```
Comment (mengxr) [inline databricks_cli/libraries/cli.py]: * `repo` -> `repository`
* If provided, do we still search Maven Central and Spark Packages?

File: databricks_cli/libraries/cli.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+INSTALL_OPTIONS = ['jar', 'egg', 'maven-coordinates', 'pypi-package', 'cran-package']
+JAR_HELP = 'JAR uploaded to dbfs or s3.'
+EGG_HELP = 'Egg uploaded to dbfs or s3.'
+MAVEN_COORDINATES_HELP = """
+Maven coordinates in the form of GroupId:ArtifactId:Version (i.e. org.jsoup:jsoup:1.7.2).
+"""
+MAVEN_REPO_HELP = """
+Maven repo to install the Maven package from. If omitted, both Maven 
+Repository and Spark Packages are searched.
```
Comment (mengxr) [inline databricks_cli/libraries/cli.py]: Should say `Maven Central Repository`.

File: databricks_cli/libraries/cli.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+"""
+MAVEN_EXCLUSION_HELP = """
+List of dependences to exclude. For example: ["slf4j:slf4j", "*:hadoop-client"].
+"""
+PYPI_PACKAGE_HELP = """
+The name of the pypi package to install. An optional exact version
```
Comment (mengxr) [inline databricks_cli/libraries/cli.py]: minor: `PyPI`

File: databricks_cli/libraries/cli.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+CRAN_PACKAGE_HELP = """
+The name of the CRAN package to install.
+"""
+CRAN_REPO_HELP = """
+The repository where the package can be found. If not specified, the default CRAN repo is used.
+"""
+
+
+@click.command(context_settings=CONTEXT_SETTINGS,
+               short_help='Install libraries on a cluster.')
```
Comment (mengxr) [inline databricks_cli/libraries/cli.py]: `Install libraries` -> `Install a library`?

File: databricks_cli/libraries/cli.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+EGG_HELP = 'Egg uploaded to dbfs or s3.'
+MAVEN_COORDINATES_HELP = """
+Maven coordinates in the form of GroupId:ArtifactId:Version (i.e. org.jsoup:jsoup:1.7.2).
+"""
+MAVEN_REPO_HELP = """
+Maven repo to install the Maven package from. If omitted, both Maven 
+Repository and Spark Packages are searched.
+"""
+MAVEN_EXCLUSION_HELP = """
+List of dependences to exclude. For example: ["slf4j:slf4j", "*:hadoop-client"].
```
Comment (mengxr) [inline databricks_cli/libraries/cli.py]: The example should be `--maven-exclusion "slf4j:slf4j" --maven-exclusion "*:hadoop-client"`

File: databricks_cli/libraries/cli.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+@require_config
+@eat_exceptions # noqa
+def install_cli(cluster_id, jar, egg, maven_coordinates, maven_repo, maven_exclusion, pypi_package, # noqa
+                pypi_repo, cran_package, cran_repo):
+    """
+    Install libraries on a cluster. Libraries must be first uploaded to dbfs or s3
+    (see `dbfs cp -h`). Unlike the API, only one library can be installed for each execution of
+    `databricks libraries install`.
+
+    Only provide one of [--jar, --egg, --maven-coordinates, --pypi-package, --cran-package].
```
Comment (mengxr) [inline databricks_cli/libraries/cli.py]: "Users should only provide ...`

File: databricks_cli/libraries/cli.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+        if cran_repo is not None:
+            cran_library['cran']['repo'] = cran_repo
+        libraries.append(cran_library)
+    assert len(libraries) == 1, 'Should have one library to install.'
+    install_libraries(cluster_id, libraries)
+
+
+def _uninstall_cli_exit_help(cluster_id):
+    click.echo("WARNING: Uninstalling libraries requires a cluster restart.")
+    click.echo("databricks clusters restart --cluster-id {}".format(cluster_id))
```
Comment (mengxr) [inline databricks_cli/libraries/cli.py]: Make it clear that this is a command user can run manually.

File: databricks_cli/libraries/cli.py [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+@click.option('--cluster-id', required=True, type=ClusterIdClickType(),
+              help=ClusterIdClickType.help)
+@click.option('--all', is_flag=True, default=False, help='If set, uninstall all libraries.')
+@click.option('--jar', cls=MutuallyExclusiveOption, mutually_exclusive=INSTALL_OPTIONS,
+              help=JAR_HELP)
+@click.option('--egg', cls=MutuallyExclusiveOption, mutually_exclusive=INSTALL_OPTIONS,
+              help=EGG_HELP)
+@click.option('--maven-coordinates', cls=MutuallyExclusiveOption,
+              mutually_exclusive=INSTALL_OPTIONS, help=MAVEN_COORDINATES_HELP)
+@click.option('--maven-repo', help=MAVEN_REPO_HELP)
```
Comment (mengxr) [inline databricks_cli/libraries/cli.py]: Is it needed? What happens if I installed one library with non-default repo and then tried uninstalling it only using the coordinates?

File: databricks_cli/libraries/cli.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+@eat_exceptions # noqa
+def uninstall_cli(cluster_id, all, jar, egg, maven_coordinates, maven_repo, maven_exclusion, # noqa
+                  pypi_package, pypi_repo, cran_package, cran_repo):
+    """
+    Mark libraries on a cluster to be uninstalled. Libraries which are marked to be uninstalled
+    will stay attached until the cluster is restarted. (see `databricks clusters restart -h`).
+    """
+    if all:
+        library_statuses = _cluster_status(cluster_id).get('library_statuses', [])
+        libraries = [l_status['library'] for l_status in library_statuses]
```
Comment (mengxr) [inline databricks_cli/libraries/cli.py]: Do we want to filter by LibraryInstallStatus? Curious what users see if this tries to uninstall a library that is in "FAILED" state.

File: databricks_cli/libraries/cli.py [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+        uninstall_libraries(cluster_id, libraries)
+        _uninstall_cli_exit_help(cluster_id)
+        return
+
+    libraries = []
```
Comment (mengxr) [inline databricks_cli/libraries/cli.py]: This is duplicate code. Should do some refactoring.

File: databricks_cli/libraries/cli.py [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+        if cran_repo is not None:
+            cran_library['cran']['repo'] = cran_repo
+        libraries.append(cran_library)
+    assert len(libraries) == 1, 'Should have one library to uninstall.'
+    uninstall_libraries(cluster_id, libraries)
+    _uninstall_cli_exit_help(cluster_id)
+
+
+@click.group(context_settings=CONTEXT_SETTINGS,
+             short_help='Utility to interact with jobs.')
```
Comment (mengxr) [inline databricks_cli/libraries/cli.py]: `with libraries`

File: tests/libraries/test_cli.py [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+ALL_CLUSTER_STATUSES_RETURN = {
+    'statuses': [{
+        'library_statuses': [{
+            'status': 'INSTALLED',
+            'is_library_for_all_clusters': False,
+            'library': {
+                'jar': 'dbfs:/test.jar'
+            }
+        }],
+        'cluster_id': '0213-212348-veeps379'
```
Comment (mengxr) [inline tests/libraries/test_cli.py]: `TEST_CLUSTER_ID`

File: tests/utils.py [hunk 13]
```diff
@@ -31,3 +31,7 @@ def wrapper(test, *args, **kwargs):
         conf.overwrite()
         return test(*args, **kwargs)
     return decorator.decorator(wrapper, test)
+
+
+def output_equals(actual, expected):
```
Comment (mengxr) [inline tests/utils.py]: * The method doesn't do what it claims. We can call it `assert_cli_output` and document the expected behavior.
* It is better to combine assert and output_equals into a single method so you can output the actual and expected if they don't match. Currently, you only get a boolean output.

File: databricks_cli/libraries/cli.py (hunk 1 above)
Comment (andrewmchen) [reply databricks_cli/libraries/cli.py]: Not sure, I copied these docs from the API documentation.

File: databricks_cli/libraries/cli.py (hunk 2 above)
Comment (andrewmchen) [reply databricks_cli/libraries/cli.py]: ditto: I copied these docs from the API documentation.

File: databricks_cli/libraries/cli.py (hunk 7 above)
Comment (andrewmchen) [reply databricks_cli/libraries/cli.py]: I've made the warning red to make the second echo pop out more.

File: databricks_cli/libraries/cli.py (hunk 8 above)
Comment (andrewmchen) [reply databricks_cli/libraries/cli.py]: I just checked. Unfortunately it is needed.

File: databricks_cli/libraries/cli.py (hunk 9 above)
Comment (andrewmchen) [reply databricks_cli/libraries/cli.py]: I don't think anything will happen, like the documentation says `Uninstalling libraries that are not installed on the cluster will have no impact but is not an error.`

**Comment by andrewmchen** [comment]
I've checked the `uninstall requires maven-exclusion` concerns you had. Unfortunately, we need to include all options like `--maven-repo` and `--maven-exclusion` to uninstall.

I still have to check the gist license issue with Justin.

File: tests/libraries/test_cli.py [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+
+CLUSTER_STATUS_RETURN = {
+    'library_statuses': [{
+        'status': 'INSTALLED',
+        'is_library_for_all_clusters': False,
+        'library': {
+            'jar': 'dbfs:/test.jar',
+        }
+    }],
+    'cluster_id': '0213-212348-veeps379'
```
Comment (mengxr) [inline tests/libraries/test_cli.py]: nit: TEST_CLUSTER_ID

**Pull request merged**